### PR TITLE
[Ready]Tesla balance and code cleanup

### PR DESCRIPTION
**Pull request by MrsTonedOne**
Muti-bolt is back, but the other ones don't go out as far and spread out more randomly

Power given has been cut down massively.

Ball expansion is now exponential. (32,64,128,256,512,1024,2048,etc)

Tesla now actually consumes energy and shrinks, causing the ball count to lower.

~~Fixes it repeatedly resetting shifting it pixel x/y causing flicks of the tesla ball apperenting 1 tile to the lowerleft~~ removing this fix, beam code would have to be redone.

Fixes it growing even if it wasn't powered by a pa

:cl:
tweak: Tesla balance changes  
tweak: Tesla will now require increasingly more energy to trigger new orbiting balls. First ball comes with 32 energy (down from 300), fourth requires 256 (down from 1200), and so on and on, doubling with each new ball. Balls after 7 are increasingly harder to get than before, whereas balls before 6 are easier to get.  
add: MULTIBOLT IS BACK!  
tweak: Be warned that in multibolt all but 1 primary bolt now have a random shocking range, meaning they may still target a close human over a far away grounding rod as the rod may be out of range.
tweak: Power output of tesla massively dropped, it should no longer put out almost twice the power of a stage 3 singulo with maxed collectors without any balls, you'll need at least 3 balls for that now. 
tweak: Tesla can now lose power, causing ball count to drop. Tesla will never shrink out of existence, just lose its balls.  
tweak: Tesla is now more likely to head in the direction of the thing it last zapped.  
/:cl:

### CODE REVIEW

**Comment by PKPenguin321** [comment]
>Ball expansion is now exponential. (100,200,400,800,1600,3200,6400,12800,etc)

does this affect performance significantly? we don't want laguloth 2

**Comment by Iamgoofball** [comment]
This is fine performance wise, we just need to balance this shit and make sure it doesn't make the engine TOO weak

**Comment by MrStonedOne** [comment]
@PKPenguin321 other kind, the amount of matter fed by the PA required to make a new ball is exponential. so 100 matter for the first ball, 200 for the second, 400 for the third, 800 for the forth. 

From it being a flat 300 each ball before.

**Comment by KazeEspada** [comment]
How does the power output compare to a level 3 singuloth(the most common singuloth?).  Is it still unable to break itself free if the PA is left on?

**Comment by PKPenguin321** [comment]
it's good then :+1:

**Comment by Lumipharon** [comment]
In terms of how much power it creates, what is it compared to the singulo? Because if it's still significantly higher you can very easily make door shocks insta kills/nuke powersinks in 5 seconds.

**Comment by Zilenan91** [comment]
@MrStonedOne Is how the Tesla moves still the same? Aka the more balls the more tiles it can teleport? And since the number of balls are drastically decreased, if the Tesla ever gets released wouldn't it barely manage to move out of containment before dying completely because it has no movement range at all?

My own thoughts on this are that we should just reduce the range down to only a few tiles. No bullshit deaths, no need to make an off-screen hearing fix that would probably be hacky as hell so you know it's coming.

**Comment by MrStonedOne** [comment]
the range is the same, and the telsa will still never die, it just loses balls until they run out (constructing movement to 2 tiles)

**Comment by Zilenan91** [comment]
How much power does it drain from APCs and the like so we know how fast it can grow if it gets that small?

**Comment by MrStonedOne** [comment]
As for power usage, I still need to run some live tests, but power usage should be above the amount engineering's smeses output even at no balls.

**Comment by MrStonedOne** [comment]
Power drained from apcs? hmmm, I don't actually know about that, I'll take a look.

**Comment by Fox-McCloud** [comment]
As I said eslewhere; multibolt is horrendously expensive on the server, I do not recommend utilizing it, at all; there's just too many effects being created and deleted in a short order---even when they GC 100% of the time, it's still very expensive.

**Comment by MrStonedOne** [comment]
multibolt was expensive when there was 40 to 100 bolts, the most you could do now is much lower than that.

**Comment by MrStonedOne** [comment]
also not every ball will lead to a zapping, as the miniballs all have a random range of 1 to 7, leaving some without anything to zap

**Comment by Shadowlight213** [comment]
Tesla movement really needs a buff to counteract losing energy. It already hangs around engineering a lot if it gets loose with only a few balls, and losing energy will make it actually getting out into the station even less likely.

File: code/modules/power/tesla/energy_ball.dm [hunk 1]
```diff
@@ -28,30 +28,44 @@ var/list/blacklisted_tesla_types = list(/obj/machinery/atmospherics,
...
+	
 	var/list/orbiting_balls = list()
 	var/produced_power
-	var/is_orbiting
+	var/energy_to_raise = 100
+	var/energy_to_lower = -10
 
 /obj/singularity/energy_ball/Destroy()
-	for(var/obj/singularity/energy_ball/EB in orbiting_balls)
+	if (orbiting && istype(orbiting, /obj/singularity/energy_ball))
```
Comment (neersighted) [inline code/modules/power/tesla/energy_ball.dm]: space

File: code/modules/power/tesla/energy_ball.dm [hunk 2]
```diff
@@ -91,29 +116,36 @@ var/list/blacklisted_tesla_types = list(/obj/machinery/atmospherics,
...
 		C.dust()
 	return
 
-/proc/get_closest_atom(type, list, source)
+/proc/get_closest_atom(list, source, typetocheck)
 	var/closest_atom
 	var/closest_distance
 	for(var/A in list)
-		if(!istype(A, type))
+		if (typetocheck && !istype(A, typetocheck))
```
Comment (neersighted) [inline code/modules/power/tesla/energy_ball.dm]: space

File: code/modules/power/tesla/energy_ball.dm [hunk 3]
```diff
@@ -200,11 +222,16 @@ var/list/blacklisted_tesla_types = list(/obj/machinery/atmospherics,
...
 					else
 						M.ex_act(1)
-			return
+			
 	if(!closest_atom)
 		closest_atom = get_closest_atom(/obj/structure, potential_structure_zaps, source)
 		if(closest_atom)
 			var/obj/structure/S = closest_atom
-			source.Beam(S,icon_state="lightning[rand(1,12)]",icon='icons/effects/effects.dmi',time=5)
+			source.Beam(S, icon_state="lightning[rand(1,12)]", icon='icons/effects/effects.dmi', time=5)
```
Comment (neersighted) [inline code/modules/power/tesla/energy_ball.dm]: All these beam copypastas could be consolidated

**Comment by MrStonedOne** [comment]
I'm still working on this,

I've been profiling modifications to tesla_zap to try to make it sanic fast.

**Comment by MrStonedOne** [comment]
I tested this by releasing it loose on the station, and everything seems to be working and no runtimes generated.

**Pull request merged**